### ISSUE TITLE
Notebooks

### ISSUE THREAD

**Issue by hbredin** [pyannote/pyannote-audio#48]
This is meant to keep track of notebooks that we may publicize in the tutorial section at some point.

- [ ] [TristouNet-based speaker change detection](https://gist.github.com/hbredin/3ef5e1307e3204b77c9ea957e528ce56)

**Comment by EngineerReversed** [comment]
Hello bredin!
You have done a pretty good work on video as well as audio repositories of pyannote. Kudos to that!
I was trying to implement the TristouNet-based speaker change detection notebook which you attached here. However, it shows me the following error while loading the model

AttributeError                            Traceback (most recent call last)

    best_epoch = 986
    pretrained_model = SequenceEmbedding.load('./pyannote-audio/tutorials/speaker-embedding/2+0.5/TristouNet/train/Etape.SpeakerDiarization.TV.train/weights/0986.h5',best_epoch)

AttributeError: type object 'SequenceEmbedding' has no attribute 'load'

I believe this error arises due to the fact that there is not 'load' model written in SequenceEmbedding class. So, I am unable to load the keras model and proceed further. Can you throw some light on it?

PS: I am using pyannote.audio==0.3 version

**Comment by hbredin** [comment]
You may want to try using the `develop` branch and update the notebook according to:
https://github.com/pyannote/pyannote-audio/blob/22862234aada1d554fa312db501ff47ddc4fc102/pyannote/audio/embedding/extraction.py#L59-L64

I'd be happy to merge a pull request fixing this (old) notebook.

**Comment by EngineerReversed** [comment]
Thanks once again for guiding me forward to the ```develop``` branch. So, I cloned the repo and switched to ```develop``` branch. Then, I placed the ```audio``` folder inside ```pyannote``` (at /env/bin/....site-packages/pyannote/) and installed dependencies accordingly to the ```develop``` branch.
I followed the above code snippet and was able to load the model successfully. MFCC features were successfully extracted and ```sequence_embedding``` object was generated.
However, when I tried executing the following command:
> embeddings = sequence_embedding.apply(first_test_file)

I received a reshape error. I am posting the stack trace below
> Traceback (most recent call last):
  File "/home/abhishek/anaconda3/envs/pyannote-audio-develop/lib/python3.5/site-packages/theano/compile/function_module.py", line 903, in __call__
    self.fn() if output_subset is None else\
ValueError: cannot reshape array of size 5120 into shape (100,16)

> During handling of the above exception, another exception occurred:

> Traceback (most recent call last):
  File "/home/abhishek/AthenasOwl/Voice_Recognition/Speaker_diarization/pyannote-audio/tristounet_embedding_temp.py", line 20, in <module>
    for window, embedding in sequence_embedding.apply(first_test_file):
  File "/home/abhishek/anaconda3/envs/pyannote-audio-develop/lib/python3.5/site-packages/pyannote/audio/embedding/extraction.py", line 160, in apply
    incomplete=True)])
  File "/home/abhishek/anaconda3/envs/pyannote-audio-develop/lib/python3.5/site-packages/pyannote/audio/embedding/extraction.py", line 159, in <listcomp>
    [batch for batch in self.from_file(current_file,
  File "/home/abhishek/anaconda3/envs/pyannote-audio-develop/lib/python3.5/site-packages/pyannote/generators/batch.py", line 379, in from_file
    incomplete=incomplete):
  File "/home/abhishek/anaconda3/envs/pyannote-audio-develop/lib/python3.5/site-packages/pyannote/generators/batch.py", line 454, in __call__
    yield self.postprocess(batch)
  File "/home/abhishek/anaconda3/envs/pyannote-audio-develop/lib/python3.5/site-packages/pyannote/generators/batch.py", line 293, in postprocess
    return self._postprocess(batch, self.signature())
  File "/home/abhishek/anaconda3/envs/pyannote-audio-develop/lib/python3.5/site-packages/pyannote/generators/batch.py", line 290, in _postprocess
    return postprocess_func(batch)
  File "/home/abhishek/anaconda3/envs/pyannote-audio-develop/lib/python3.5/site-packages/pyannote/audio/embedding/extraction.py", line 131, in postprocess_ndarray
    return self.embed_(X)
  File "/home/abhishek/anaconda3/envs/pyannote-audio-develop/lib/python3.5/site-packages/pyannote/audio/embedding/extraction.py", line 106, in embed
    return K_func([batch, 0])[0]
  File "/home/abhishek/anaconda3/envs/pyannote-audio-develop/lib/python3.5/site-packages/keras/backend/theano_backend.py", line 1222, in __call__
    return self.function(*inputs)
  File "/home/abhishek/anaconda3/envs/pyannote-audio-develop/lib/python3.5/site-packages/theano/compile/function_module.py", line 917, in __call__
    storage_map=getattr(self.fn, 'storage_map', None))
  File "/home/abhishek/anaconda3/envs/pyannote-audio-develop/lib/python3.5/site-packages/theano/gof/link.py", line 325, in raise_with_op
    reraise(exc_type, exc_value, exc_trace)
  File "/home/abhishek/anaconda3/envs/pyannote-audio-develop/lib/python3.5/site-packages/six.py", line 685, in reraise
    raise value.with_traceback(tb)
  File "/home/abhishek/anaconda3/envs/pyannote-audio-develop/lib/python3.5/site-packages/theano/compile/function_module.py", line 903, in __call__
    self.fn() if output_subset is None else\
ValueError: cannot reshape array of size 5120 into shape (100,16)
Apply node that caused the error: Reshape{3}(Elemwise{Add}[(0, 0)].0, TensorConstant{[ -1 100  16]})
Toposort index: 129
Inputs types: [TensorType(float32, matrix), TensorType(int64, vector)]
Inputs shapes: [(320, 16), (3,)]
Inputs strides: [(64, 4), (8,)]
Inputs values: ['not shown', array([ -1, 100,  16])]
Outputs clients: [[Join(TensorConstant{2}, Reshape{3}.0, Reshape{3}.0, Reshape{3}.0, Reshape{3}.0)]]

> Backtrace when the node is created(use Theano flag traceback.limit=N to make it longer):
  File "/home/abhishek/anaconda3/envs/pyannote-audio-develop/lib/python3.5/site-packages/keras/engine/topology.py", line 2497, in from_config
    process_node(layer, node_data)
  File "/home/abhishek/anaconda3/envs/pyannote-audio-develop/lib/python3.5/site-packages/keras/engine/topology.py", line 2454, in process_node
    layer(input_tensors[0], **kwargs)
  File "/home/abhishek/anaconda3/envs/pyannote-audio-develop/lib/python3.5/site-packages/keras/engine/topology.py", line 602, in __call__
    output = self.call(inputs, **kwargs)
  File "/home/abhishek/anaconda3/envs/pyannote-audio-develop/lib/python3.5/site-packages/keras/layers/wrappers.py", line 294, in call
    y = self.forward_layer.call(inputs, **kwargs)
  File "/home/abhishek/anaconda3/envs/pyannote-audio-develop/lib/python3.5/site-packages/keras/layers/recurrent.py", line 340, in call
    preprocessed_input = self.preprocess_input(inputs, training=None)
  File "/home/abhishek/anaconda3/envs/pyannote-audio-develop/lib/python3.5/site-packages/keras/layers/recurrent.py", line 1084, in preprocess_input
    timesteps, training=training)
  File "/home/abhishek/anaconda3/envs/pyannote-audio-develop/lib/python3.5/site-packages/keras/layers/recurrent.py", line 58, in _time_distributed_dense
    x = K.reshape(x, (-1, timesteps, output_dim))
  File "/home/abhishek/anaconda3/envs/pyannote-audio-develop/lib/python3.5/site-packages/keras/backend/theano_backend.py", line 841, in reshape
    y = T.reshape(x, shape)

> HINT: Use the Theano flag 'exception_verbosity=high' for a debugprint and storage map footprint of this apply node.

I am trying to understand the error but it looks like theano stack trace which most probably is occuring due to incompatible dimensions of layers. Earlier I thought it is due to the ```duration``` parameter but it shows same error for different values of duration. 

PS: I am currently using ```theano==1.0.1``` and ```keras==2.0.8```

**Comment by hbredin** [comment]
Can you please open a new issue with detailed instructions on how to reproduce the error?
"New issue" because we kind of lost the initial purpose of this issue (keeping track of sample notebooks).
"Detailed instructions" because, without this, I cannot really help yoU.

**Comment by EngineerReversed** [comment]
@hbredin 
The issue is related to this the current notebook. As I figured out later on if you keep the `duration=3.2`, you would get the above error. However, if you keep the `duration=2.0`, the code would execute successfully. It turns out that the `duration` is dependent on the input shape of the network.

PS: I did some fixes as per the latest development in `develop` branch and was able to successfully execute the above notebook. I am attaching the new notebook.
[TristouNet.SCD.ipynb.zip](https://github.com/pyannote/pyannote-audio/files/1663361/TristouNet.SCD.ipynb.zip)
